code/modules/hydroponics/seed_extractor.dm

```diff
@@ -107,107 +113,129 @@
-		dat += "</table>"
-	var/datum/browser/popup = new(user, "seed_ext", name, 700, 400)
-	popup.set_content(dat)
-	popup.open()
-	return
+	switch(action)
+		if("vend")
+			vendSeed(text2num(params["seedid"]), vend_amount)
+			add_fingerprint(usr)
+			SStgui.update_uis(src)
+		if("set_vend_amount")
+			if(!length(params["vend_amount"]))
+				return
+			vend_amount = clamp(text2num(params["vend_amount"]), 1, MAX_DISPENSE_SEEDS)
+			add_fingerprint(usr)
 
-/obj/machinery/seed_extractor/Topic(href, list/href_list)
-	if(..())
-		return 1
-	usr.set_machine(src)
-
-	href_list["li"] = text2num(href_list["li"])
-	href_list["en"] = text2num(href_list["en"])
-	href_list["ma"] = text2num(href_list["ma"])
-	href_list["pr"] = text2num(href_list["pr"])
-	href_list["yi"] = text2num(href_list["yi"])
-	href_list["pot"] = text2num(href_list["pot"])
-
-	for (var/datum/seed_pile/N in piles)//Find the pile we need to reduce...
-		if (href_list["name"] == N.name && href_list["li"] == N.lifespan && href_list["en"] == N.endurance && href_list["ma"] == N.maturation && href_list["pr"] == N.production && href_list["yi"] == N.yield && href_list["pot"] == N.potency)
+/obj/machinery/seed_extractor/proc/vendSeed(seed_id, amount)
+	if(!seed_id)
+		return
+	var/datum/seed_pile/selected_pile
+	for(var/datum/seed_pile/N in piles)
+		if(N.id == seed_id)
+			amount = clamp(amount, 0, N.amount)
+			N.amount -= amount
+			selected_pile = N
 			if(N.amount <= 0)
-				return
-			N.amount = max(N.amount - 1, 0)
-			if (N.amount <= 0)
 				piles -= N
-				qdel(N)
 			break
-
-	for (var/obj/T in contents)//Now we find the seed we need to vend
-		var/obj/item/seeds/O = T
-		if (O.plantname == href_list["name"] && O.lifespan == href_list["li"] && O.endurance == href_list["en"] && O.maturation == href_list["ma"] && O.production == href_list["pr"] && O.yield == href_list["yi"] && O.potency == href_list["pot"])
-			O.forceMove(loc)
+	if(!selected_pile)
+		return
+	var/amount_dispensed = 0
+	for(var/obj/item/seeds/O in contents)
+		if(amount_dispensed >= amount)
 			break
-
-	updateUsrDialog()
-	return
+		if (O.plantname == selected_pile.name && O.lifespan == selected_pile.lifespan && O.endurance == selected_pile.endurance && O.maturation == selected_pile.maturation && O.production == selected_pile.production && O.yield == selected_pile.yield && O.potency == selected_pile.potency)
+			O.forceMove(loc)
+			amount_dispensed++
 
 /obj/machinery/seed_extractor/proc/add_seed(obj/item/seeds/O)
-	if(contents.len >= 999)
-		to_chat(usr, "<span class='notice'>\The [src] is full.</span>")
-		return 0
+	if(!O || !ishuman(usr) || !Adjacent(usr))
+		return
+	if(length(contents) >= max_seeds)
+		to_chat(usr, "<span class='notice'>[src] is full.</span>")
+		return
 
-	if(istype(O.loc,/mob))
+	if(ismob(O.loc))
 		var/mob/M = O.loc
-		if(!M.drop_item())
-			return 0
-	else if(istype(O.loc,/obj/item/storage))
+		M.drop_item()
```

Comment:
Pretty sure you still need the `if(!M.drop_item()` because that checks for `NODROP`.

Reply:
added